### ISSUE TITLE
Lecture ten- SVM, Performance evaluation

### ISSUE THREAD

**Issue by jimjinyan** [jimjinyan/Go-Just-Go#24]
## SVM

SVM start from perceptron which is the seperating hyperplane
### Separting Hyperplane
- $\cal{W}^Tx =0$
    - based on the dot product, if the dot product is 0, then these two vector are vertical
    - The hyperplane are not exits in our data, it is just a plane we image
    - there are many vectors in x_1 and x_2 plane being verical with $\cal{W}$
    - we are just looking for a specail plane which can sperate all data poins porperly
    -  for $\cal{W}^Tx =0$, if $W$ are not zero, then the x have to be zero, so we need add interception
-  So the hyperplane function is in the from: - $\cal{W}^Tx +b =0$
- Therefore the SVM model are decided by several paramenters like $\cal{W}$ and certain points.
    -  put points in anove fucntion to see if big or less than 0, then give corresponding label
        - The $cal{W}$must be decided to so, but how?
            - in 50's, it seems the  $cal{W}$ are decided by training data.
    - Advange comparing with KNN
    - Disadvantage: liearity for the separtig Hyperplane
- Perceptron
    - The whole process like a neuron
- for separting hyperplane seems diffcult do solve the XOR Problem
    -  it just need to transform the data to like higher dimension so the data are seperated (like the ideal of kernal in SVM for non-linear)
- Sigmoid Activation
    -  it is wildly used in like Logistic Regression. 
    -  [A good refference](https://blog.csdn.net/wolfblood_zzx/article/details/74453434)

### Support Vector Machine
-  using the concerpt Maximum Margin Classification
    - Supoort vectors who decide the seprating hyperplan and the maximum margin
    -  still need certian amount of data even though the few support vectors are only concerns. The reason is with more data, we have higher probabiblity to find the right support vectors.
- Maximum Margin Classification--the way to decide the hyperplane(not like the Perceptron)
    - *Reffering the other materials for SVM study*
- for Maximum Margin Classification, all the model care about is the data who is close to the hyperplane
   -  The model do care about a good apple but care about the apple likes orange (rare cases)
-  the support vectors are very senstive for the mislabeled or outliers- so we introduced the slack variables
- slack variables
    - note, it is distance to its marigin lie the opposite side on (not the seperating hyperplane)
    -  The "C" in sklearn
- here I felt confusion since we compar the hyperplane for getting a right label and then we add slack variables for some outliers and mislabeled, then what the margin do for the SVM? The data seems do not compare with those mrgin?
    - get this confusion is becase I do not really under the SVM. The important concept is with or without marign , the hyperplane will be different! Without marign, the hyperplane will be like perceptron in 50's. Becase of the margin, we got a new hyperplane. If we comparing data with this hyperplane, it might case some mislabel, but it give us the most robust hyperplane. Plus, how many mislable are contorled by paramenter C  

- How does SVM solve the XOR Problem
    - similar, we add more dimension to the data set. So in higher diension spance, we still juse linear seperating hyperplan while its project on the original dimension is non-linear hyperplan. 

**Math for SMV**
- There are multiple *separating hyperplanes* which can be decide by $\theta^T\cal{x}=0$. How to find the best separating hyperplanes?
    - Insted of using probabiity in logistic regression, the distance is used
    - To explicitly treat the intercept, term $b$ and$\cal{w}$ are used$$h_{\cal{w}, \cal{b}}(x)=g(\cal{w}^T\cal{x}+b).$$If $\cal{z}=\cal{w}^T\cal{x}+b$, then $g(\cal{z})=1$ if $\cal{z}\ge0$, and $g(\cal{z})=-1$ if $\cal{z}\<0$
    - **functional margin of ($\cal{w},\ $\cal{b})**$$\hat{\gamma}^{(\cal{i})}=\cal{y}^{(\cal{i})}(\cal{w}^T\cal{x}+\cal{b}).$$ If $\cal{y}^{(\cal{i})}(\cal{w}^T\cal{x}+\cal{b})>0$, the our prediction is correct, so it represents a confident and a correct prediction.
        - Given a training set, the smallest of the functional margins is $$\hat{\gamma}=min_{i =1,2...,m}\hat{\gamma}^{(\cal{i})}$$
    - **geometric margins**$$\gamma^{(i)}=\left( \left(\frac{\cal{w}}{\rVert\cal{w}\rVert}\right)^Tx^{(i)} + \frac{b}{\rVert\cal{w}\rVert}\right).$$ When $\rVert\cal{w}\rVert=1$, the functional margin equals to geometric margin
        - Note there are a difference between $\gamma$ for geometric margins and functional margin
        - *note*: the dot multiply of two vectors is 0 means the are vertical. But for vector, they always starts from 0, so in the separating hyperplanes,it means we need to move $-b$ back
    -  optimization problem:$$max_{\gamma, \cal{w}, \cal{b}}\ \ \ \gamma\ \ \ \ s.t. \cal{y}^{(\cal{i})}(\cal{w}^T\cal{x}+\cal{b})\ge\gamma; \ \ \ \cal{i}=1,...,m;\ \ \  \rVert\cal{w}\rVert=1$$
        - the key during the optimization is to keep doing some transform, finding the a conve quadratic objective.
            - by maximize $\frac{\hat{\gamma}}{\rVert\cal{w}\rVert}$, it make the functional margins and geometric margins are equal and get fird of one constrains of $\rVert\cal{w}\rVert=1$
            - Since the $w,\ b$ can be arbitrary scaling, we scale them to get $\hat{\gamma}=1$
            - maximize $\frac{\hat{\gamma}}{\rVert\cal{w}\rVert}$ is simiare as minimize $\frac{1}{2}\rVert\cal{w}\rVert^2$
        - Then using Lagrange duality and KKT to solve the $w,\ b$
            - The SVM happens to satisfy the KKT condition
            - KKT dual complementarity condition shows except support vectors, other vectors's $a_i=0$, so they do not attend the whole SVM calculation
            - by Lagrange and KKT, the $w,\ b$ have been represented by $a_i$, so by dual optimization, we can only maxmize $a_i$. After get $a_i$, then we can get $w$, after get$w$, we can get $b$
    - How the $w$ decided and the ideal about Kernels
        - starts with equation which get from Lagrange$$\cal{w} = \sum_{i=1}{m}a_iy^{(i)x^{(i)}}$$For SVM, $x$ is the vector with feactures(can be as many as it), $y$ is the label. Therefore it is the training data to get the $\cal{w}$ which is decided by maxmized $a$ (as mentiond above).
        - Start to use the $\cal{w}$ to do predction on the testing data, which will be:$$\cal{w}^Tx+b$$Note, here $x$is the feacture vector of testing data. Then we show $\cal{w}$ as we got by training data, it will be$$\left( \sum_{i=1}^ma_iy^{(i)}x_{support vector}^{(i)}\right)^Tx_{test}+b$$The key is 1) m is the quantity of support vector and the useage of training x and testing x. Do a slightly transform, we got$$\sum_{i=1}^ma_iy^{(i)}\left( x_{support vector}^{(i)}\right)^Tx_{test}+b$$This perfect solved the linear SVM.

- Kernels
Kernels are used to answer how about non-linear SVM.
**Given a feature mapping $\Phi$, we define the corresponding Kernel to be$$K(x,\ z)=\Phi(x)^T\Phi(z)$$**Here is basically saying that Kernel is kind of dot multiply for mapping of support vector and testing data
- The problem is it is too expensive to calculate in most of time, how to do?  The answer is Kernels
- If all K functon can be a valid kernel, which means can be like $K(x,\ z)=\Phi(x)^T\Phi(z)$ (can we find $\Phi$)?
    - Hence, we’ve shown that if K is a valid kernel (i.e., if it corresponds tosome feature mapping φ), then the corresponding Kernel matrix K ∈ Rm×m is symmetric positive semidefinite. (also called a Mercer kernel)   

- Slack Variable
    - for non-separable case
    - Slack variable is also like support vectors 
    - each outlier has its own slack variable. 

- Kernel Trick for SVMs
    - Arbitrar many dimensions with little computational cost
    - Maximal margin helps with curse of dimensionality--becase bigger margin can be more general(reduce the overfit)

- Radial Basis function(RBF):
    - two parameters to tune: $\gamma$ and C
        - The biger the two paramenters, it is easier for overfit

**Tips and tricks for SVM**
1. Normalize data!!
    - mean: 0, std:1
    - map to [0,1] or [-1,1]
2. Normalize test set in the same way!
    - test data can not be used to estimate the mean and std for training data
3. Always try RBF kernel
4. SVM need less training data during prediction comparing with KNN
5. SVM has more parameters to tune, like kernal, C, $\gamma$, while KNN only have parameter K

**Error Measures**
- The diagonal is the right while off-diagonal are mistakes
- ROC(Reciever Operating Characteristic) is using true positive rate and false postive rate
    - The threshold is something like probablity, distance to hyperplane, etc
    - Each parameter will have one curve, which use area under curve to show it is good or bed
- Precision Recall Curve
    - The recall(true positive rate) is biger and biger, which means we classifer the more data as positive(no mtter true or not), so it is more like(宁可错杀3000也不放过1个). So in this case, we need to use precision to see if how much of  them are correctly classified. 
    - The good things are PRC do not care about the true negtive, which most of time they are noise. So it is useful for **un-balance data**.

- F measure

**Multi Class**
-  one vs. all
- one vs. one
